sdk/src/Core/Internal/Utils/HostEndPoint.cs

```diff
@@ -0,0 +1,168 @@
+﻿//-----------------------------------------------------------------------------
+// <copyright file="HostEndPoint.cs" company="Amazon.com">
+//      Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+//      Licensed under the Apache License, Version 2.0 (the "License").
+//      You may not use this file except in compliance with the License.
+//      A copy of the License is located at
+//
+//      http://aws.amazon.com/apache2.0
+//
+//      or in the "license" file accompanying this file. This file is distributed
+//      on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+//      express or implied. See the License for the specific language governing
+//      permissions and limitations under the License.
+// </copyright>
+//-----------------------------------------------------------------------------
+
+using System;
+using System.Linq;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading;
+using Amazon.Runtime.Internal.Util;
+
+namespace Amazon.XRay.Recorder.Core.Internal.Utils
+{
+	/// <summary>
+	/// Represents a endpoint on some network.
+	/// The represented endpoint is identified by a hostname.
+	///
+	/// Internally resolves and caches an ip for the hostname.
+	/// The ip is cached to keep the normal path speedy and non-blocking.
+	/// </summary>
+	public class HostEndPoint
+	{
+		private const int CACHE_TTL = 60;	//Seconds to consider a cached dns response valid
+		private IPEndPoint _ipCache = null;
+		private DateTime? _timestampOfLastIPCacheUpdate = null;
+		private static readonly Logger _logger = Logger.GetLogger(typeof(HostEndPoint));
+		private ReaderWriterLockSlim cacheLock = new ReaderWriterLockSlim();
+
+		/// <summary>
+		/// Create a HostEndPoint.
+		/// </summary>
+		/// <param name="host"></param>
+		/// <param name="port"></param>
+		public HostEndPoint(string host, int port)
+		{
+			Host = host;
+			Port = port;
+		}
+		
+		/// <summary>
+		/// Get the hostname that identifies the endpoint.
+		/// </summary>
+		public string Host { get; }
+		/// <summary>
+		/// Get the port of the endpoint.
+		/// </summary>
+		public int Port { get; }
+
+
+		private bool IsIPCacheValid()
+		{
+			bool entered = cacheLock.TryEnterReadLock(0);
+
+			if (!entered)
```

Comment:
I think we should not check this condition. The reason : `GetIPEndPoint()` calls this method. So suppose, 1 thread return value false, and enters `UpdateCache()` code block with the write lock and at the same time another thread hits `IsIPCacheValid()`, the current logic of returning value `false` at this point  will result in calling `UpdateCache()` unnecessarily for the second time. `UpdateCache()`is expensive call since it has write lock with DNS resolution operation. As per convention, lets have read lock in try finally block, to avoid any kind of exception  and change read lock to `EnterReadLock`

Reply:
Did som rework on HostEndPoint so that the double update no longer occurs. Also included unit tests to check for that.